### PR TITLE
Debug "CI failing for cgroup2+docker+rootless"

### PR DESCRIPTION
**Pull request by AkihiroSuda**
Debug #2409

### CODE REVIEW

**Comment by aojea** [comment]
are these restars of kubelet normal?

```
Aug 08 15:20:37 kind-control-plane systemd[1]: Starting kubelet: The Kubernetes Node Agent...
Aug 08 15:20:37 kind-control-plane sh[484]: ERROR: this script needs /sys/fs/cgroup/cgroup.procs to be empty (for writing the top-level cgroup.subtree_control)
Aug 08 15:20:37 kind-control-plane systemd[1]: kubelet.service: Control process exited, code=exited, status=1/FAILURE
Aug 08 15:20:37 kind-control-plane systemd[1]: kubelet.service: Failed with result 'exit-code'.
Aug 08 15:20:37 kind-control-plane systemd[1]: Failed to start kubelet: The Kubernetes Node Agent.
Aug 08 15:20:38 kind-control-plane systemd[1]: kubelet.service: Scheduled restart job, restart counter is at 89.
Aug 08 15:20:38 kind-control-plane systemd[1]: Stopped kubelet: The Kubernetes Node Agent.
Aug 08 15:20:38 kind-control-plane systemd[1]: Starting kubelet: The Kubernetes Node Agent...
Aug 08 15:20:38 kind-control-plane systemd[1]: Started kubelet: The Kubernetes Node Agent.
```

the podman job doesn't have them

File: pkg/cluster/internal/providers/docker/provision.go [hunk 1]
```diff
@@ -245,6 +245,8 @@ func runArgsForNode(node *config.Node, clusterIPFamily config.ClusterIPFamily, n
 		"--volume", "/lib/modules:/lib/modules:ro",
 		// propagate KIND_EXPERIMENTAL_CONTAINERD_SNAPSHOTTER to the entrypoint script
 		"-e", "KIND_EXPERIMENTAL_CONTAINERD_SNAPSHOTTER",
+		// unset container=docker and fake the value, for debugging https://github.com/kubernetes-sigs/kind/issues/2409
+		"-e", "container=podman",
```
Comment (AkihiroSuda) [inline pkg/cluster/internal/providers/docker/provision.go:249]: This seems to work, but not sure why.

cc @ktock 
https://github.com/kubernetes-sigs/kind/pull/2354 might be related?

File: pkg/cluster/internal/providers/docker/provision.go [hunk 2]
```diff
@@ -131,6 +132,7 @@ func createContainer(args []string) error {
 	if err := exec.Command("docker", args...).Run(); err != nil {
 		return errors.Wrap(err, "docker run error")
 	}
+	time.Sleep(10 * time.Second) // debug
```
Comment (AkihiroSuda) [inline pkg/cluster/internal/providers/docker/provision.go:135]: The “ ERROR: this script needs /sys/fs/cgroup/cgroup.procs to be empty” issue was happening when `docker exec -i` was executed before systemd initializes `/sys fs/cgroup`.

So this sleep works as a workaround.

Comment (aojea) [reply pkg/cluster/internal/providers/docker/provision.go:135]: do we have some way to "detect it" instead of sleeping?

Comment (BenTheElder) [reply pkg/cluster/internal/providers/docker/provision.go:135]: I guess we could start watching for a log line in the node output if this is the case.

Comment (AkihiroSuda) [reply pkg/cluster/internal/providers/docker/provision.go:135]: I'm not sure.

Grepping `Reached target .*Multi-User System.*` is fine?

Comment (BenTheElder) [reply pkg/cluster/internal/providers/docker/provision.go:135]: I guess. It's unfortunate that this is the case, we have codepaths that should be fine without a running system (e.g. catting the version file, building kubeadm config), and speed is definitely a priority even if we haven't improved on it much lately.

Maybe we can conditionally wait on this when rootless?

**Comment by BenTheElder** [comment]
Did we ever determine what happens if we roll back the image https://github.com/kubernetes-sigs/kind/issues/2409#issuecomment-894803863? Or is it some docker change and the exec timing?

Sorry, still haven't had time to poke at this one

File: pkg/cluster/internal/providers/docker/provision.go (hunk 2 above)
Comment (AkihiroSuda) [reply pkg/cluster/internal/providers/docker/provision.go:135]: Opened https://github.com/kubernetes-sigs/kind/pull/2421

> Maybe we can conditionally wait on this when rootless?

I don't think this is specific to rootless, though specific to cgroup v2.

**Comment by AkihiroSuda** [comment]
Closing this and opened #2421 

> Did we ever determine what happens if we roll back the image #2409 (comment)? Or is it some docker change and the exec timing?


I don't think this is a "regression" in the base image nor in docker.
This is a timing issue (on cgroup v2 hosts).

**Pull request closed without merging**